server/src/main/java/io/littlehorse/server/streams/lhinternalscan/publicrequests/SearchPrincipalRequestModel.java

```
@@ -0,0 +1,132 @@
+
+    @Override
+    public GetableClassEnum getObjectType() {
+        return GetableClassEnum.PRINCIPAL;
+    }
+
+    @Override
+    public LHStore getStoreType() {
+        return LHStore.GLOBAL_METADATA;
+    }
+
+    @Override
+    public TagStorageType indexTypeForSearch() throws LHApiException {
+        return TagStorageType.LOCAL;
+    }
+
+    @Override
+    public SearchScanBoundaryStrategy getScanBoundary(String searchAttributeString) throws LHApiException {
+        return new TagScanBoundaryStrategy(
+                searchAttributeString, Optional.ofNullable(earliestStart), Optional.ofNullable((latestStart)));
+    }
+
+    @Override
+    public SearchPrincipalRequest.Builder toProto() throws LHApiException {
+        SearchPrincipalRequest.Builder builder = SearchPrincipalRequest.newBuilder();
+
+        if (bookmark != null) builder.setBookmark(bookmark.toByteString());
+
+        if (limit != null) builder.setLimit(limit);
+
+        if (earliestStart != null) builder.setEarliestStart(LHUtil.fromDate(earliestStart));
+        if (latestStart != null) builder.setLatestStart(LHUtil.fromDate(latestStart));
+
+        switch (type) {
+            case ISADMIN:
+                builder.setIsAdmin(isAdmin);
+                break;
+            case TENANTID:
+                builder.setTenantId(tenantId);
+                break;
+            case PRINCIPALCRITERIA_NOT_SET:
+                throw new LHApiException(Status.FAILED_PRECONDITION, "Principal query criteria is not valid.");
+        }
+
+        return builder;
+    }
+
+    @Override
+    public void initFrom(Message proto, ExecutionContext context) throws LHSerdeError {
+        SearchPrincipalRequest p = (SearchPrincipalRequest) proto;
+
+        if (p.hasBookmark()) {
+            try {
+                this.bookmark = BookmarkPb.parseFrom(p.getBookmark());
+            } catch (Exception exn) {
+                log.error("Failed to load bookmark: {}", exn.getMessage(), exn);
```

I know this is a pattern for our other search classes, but I think we should throw `LHApiException` in this case because users can send an invalid bookmark and still receive the wrong results page. I would like to hear @coltmcnealy-lh 's opinion.

---

I agree, it should be `LHApiException` with `Status.INVALID_ARGUMENT`

---

Should we apply this change for all SearchRequestModels?

---

Yes, we should. but for now, just this class is fine. We can create a separate PR for the other search classes.